Add HNPWA-Zuix implementation

genemars opened the pull request:
Implementation using ZUIX.

genielabs commented:
@kristoferbaxter  done! =)

kristoferbaxter commented:
Awesome! I'll take a look at the lighthouse info later today.

Thank you for your hard work to make this happen!

genielabs commented:
@kristoferbaxter I made some code changes. Most just to clean draft stuff and add more comments/docs.
Do I have to rerun/update lighthouse test results? Also I noticed that if I disable the Google Fonts inclusion in the app it then gets very much nicer results =/

genielabs commented:
**with fonts:**
https://www.webpagetest.org/lighthouse.php?test=170604_72_b2164d23d46e613ecd06fc0ed6abae92&run=2

**without fonts:**
https://www.webpagetest.org/lighthouse.php?test=170604_81_14aecf4b883687605e280c5648859915&run=2

kristoferbaxter commented:
I wouldn't worry about over-optimizing the results, but instead carefully evaluate the costs of adding required fonts for end user experience.

The goal of this project isn't a pure performance comparison, instead to show many options are available to make a PWA that responds quickly to user requests.

genielabs commented:
I added the font resource to the service worker cache. Lighthouse is still reporting this as a issue, but custom fonts are used in real applications and are important in term of ux so I decided to keep it despite of the results.

housseindjirdeh commented on site/_apps/zuix-hn.txt:
```diff
@@ -0,0 +1,26 @@
+---
+weight: 2
```
Will it be okay if you changed this to `9` so it shows up at the bottom of the page? Although we only have a single landing page with scrolling to show all implementations, I think it's only fair incremental additions show up one after another :)

genielabs replied:
Didn't really know what was that for. I did copy&paste and left it as is. Of course it is ok.

housseindjirdeh replied:
:P Sorry for the ambiguity there

housseindjirdeh commented on site/_apps/zuix-hn.txt:
```diff
@@ -0,0 +1,26 @@
+    - name: 'Gene'
+lighthouse: '91/100'
+interactive-em: '3.8s'
+interactive-faster-3g: '2.6s'
```
It looks like you obtained these results from WPT` First View` values. For the rest of the implementations we've been using the times from the Lighthouse results linked through Webpagetest. You may want to do the same here and just use the `First Interactive` times linked in those reports (please let us know however if you're seeing issues getting those values)

genielabs replied:
No, this is the First interactive time as you can see from the linked WebPageTest results in this same file. Now those could be a little slower due to custom fonts inclusion.

housseindjirdeh replied:
Hmm if I'm not mistaken these are the `First View` times from WPT directly correct? Really sorry for the confusion, but what I meant is that with each Webpagetest, we make sure to generate Lighthouse results and then from there find the First Interactive times. I can't find associated LH results with your WPT tests so I'm assuming you didn't get it from there.

![image](https://cloud.githubusercontent.com/assets/12476932/26809349/02f915e8-4a32-11e7-87cd-bd96c51500c5.png)

![image](https://cloud.githubusercontent.com/assets/12476932/26809377/50d8c588-4a32-11e7-889e-09c2d15b1af5.png)

Please let me know if this makes sense. Since this seems ambigious, I really need to flesh out `CONTRIBUTING.md` and even include screenshots if necessary so I apologize if it wasn't clear enough.

genielabs replied:
Ok never mind :) I have to fix this then.

genielabs replied:
done!

housseindjirdeh commented on site/_contributors/gene.txt:
```diff
@@ -0,0 +1,6 @@
+---
+weight: 10
```
Similarly here, probably makes sense to make sure this number is 1+ the last contributor

genielabs replied:
`position` would have been more easy to decipher =) I will fix this as well.

housseindjirdeh replied:
:P :P I apologize! I'll open an issue to make sure this is changed to `position` (I agree it makes more sense) + explain it clearly in the contributing doc

genielabs replied:
Made requested changes, everything should be in place now  =)

genielabs commented on site/_apps/zuix-hn.txt:
```diff
@@ -0,0 +1,26 @@
+libraries: 
+    - name: 'ZUIX'
+module-bundling: 'zuix-bundler'
+service-worker: 'Application Shell'
```
The service worker is just implement simple caching of app files. Network requests are not being cached in this very moment. But now I see that Workbox can come handy for implementing caching in the sw. I will use that later. Thanks for the hint.

housseindjirdeh replied:
No worries at all :) But yep in the meantime this is just fine.

genielabs commented:
@housseindjirdeh is it ok now?

housseindjirdeh commented:
Looks good to me :+1: 

@addyosmani @kristoferbaxter feel free to do a quick check before we merge. Sanity check with filmstrip view seems to show WPT running correctly here but please let me know if I missed something

addyosmani commented on site/_apps/zuix-hn.txt:
```diff
@@ -14,11 +14,11 @@ hosting: 'Firebase'
+interactive-em: '3.2s'
+interactive-faster-3g: '2.2s'
+lighthouse-link: https://www.webpagetest.org/lighthouse.php?test=170611_5G_d360ab4227e495c34cfb8e6ea58b37e4&run=3
+wpt-em-link: https://www.webpagetest.org/result/170611_QG_beecb3a95d8b5e66ed6a33522b1a20a7/
```
@housseindjirdeh @kristoferbaxter There's something a little strange going on in the very latest trace over at https://www.webpagetest.org/video/compare.php?tests=170611_QG_beecb3a95d8b5e66ed6a33522b1a20a7-r:1-c:0.

I see:
- main-thread activity settling down initially after 3.7s (consistently after 7.5/8s)
- earliest the app displays interactive UI is about 6s

This isn't OP's fault at all as Lighthouse reports different numbers here. Based on what I've seen upstream, we might want to wait until latest Lighthouse (2.1.0) gets deployed to WPT and re-run the numbers then before landing.

Does that sound reasonable?

housseindjirdeh replied:
Absolutely, thanks for spotting some unusual trace activity. We can definitely wait for 2.1.0 to get deployed before we re-run the numbers.

kristoferbaxter replied:
I agree as well.

housseindjirdeh replied:
Reran tests - Lighthouse First Interactive numbers are looking more consistent now with filmstrip traces.

EM: (6.7s) -
https://www.webpagetest.org/result/170612_MY_2c31226e85ddc0fdb377bccc8bfb279a/

3G: (5.48s) -https://www.webpagetest.org/result/170612_NA_2c6280b97db6ba7b25a853b014f9bafa/

Let me know if the numbers look reasonable and I didn't miss anything and we can merge this in with updated stats.

genielabs replied:
@housseindjirdeh, I've minified the *app.bundle.js* and updated results with lighthouse 2.1.0.
EM: (5.1s) - https://www.webpagetest.org/lighthouse.php?test=170613_22_be140a755468b49f0e8480eeb79149ed&run=1
3G: (4.3s) - https://www.webpagetest.org/result/170613_Y4_3f81ed03de518aaecc6a8823e74ec401/
Anyway I got very different results by running the tests in different moments, mainly due to the network goodness.

addyosmani commented:
@genemars Feel free to re-ping us for a review once you feel you're ready. I noticed some extra commits going in and didn't want to prematurely approve if there's more work coming in :)

genemars commented:
@addyosmani please go for a review. I was thinking about writing firebase functions to wrap standard HN firebase API and implement listing with one API call instead of having client to make a call for each item in the list. But despite the preformance it is perhaps interesting to see a different implementation and leave it as is for now. Suggestions very accepted =)

genemars commented:
@addyosmani @housseindjirdeh @kristoferbaxter Can you please review/merge this? :-D

addyosmani commented:
Yep. I'll take another look now.

addyosmani commented on site/_apps/zuix-hn.txt:
```diff
@@ -6,19 +6,19 @@ library: 'ZUIX'
+interactive-faster-3g: '3.0s'
+lighthouse-link: https://www.webpagetest.org/lighthouse.php?test=170613_KA_4797d0c9289924c6d61abf3326146944&run=1
+wpt-em-link: https://www.webpagetest.org/result/170613_1E_b545fe08a95f7405985fc1e6a54254dd/
+wpt-faster-3g-link: https://www.webpagetest.org/result/170613_KA_4797d0c9289924c6d61abf3326146944/
```
This WPT link is now down. New one for faster 3G: https://www.webpagetest.org/result/170715_NN_e8f85dda8cffc693f00a914dc543bfd2/

addyosmani commented on site/_apps/zuix-hn.txt:
```diff
@@ -6,19 +6,19 @@ library: 'ZUIX'
+interactive-em: '4.8s'
+interactive-faster-3g: '3.0s'
+lighthouse-link: https://www.webpagetest.org/lighthouse.php?test=170613_KA_4797d0c9289924c6d61abf3326146944&run=1
+wpt-em-link: https://www.webpagetest.org/result/170613_1E_b545fe08a95f7405985fc1e6a54254dd/
```
EM link for WPT: https://www.webpagetest.org/result/170715_16_50fe1017f7daf7f9ca735dbe8726111c/

addyosmani commented on site/_apps/zuix-hn.txt:
```diff
@@ -6,19 +6,19 @@ library: 'ZUIX'
-wpt-faster-3g-link: https://www.webpagetest.org/result/170613_Y4_3f81ed03de518aaecc6a8823e74ec401/
+interactive-em: '4.8s'
+interactive-faster-3g: '3.0s'
+lighthouse-link: https://www.webpagetest.org/lighthouse.php?test=170613_KA_4797d0c9289924c6d61abf3326146944&run=1
```
New Lighthouse link as current one is down: https://www.webpagetest.org/lighthouse.php?test=170715_NN_e8f85dda8cffc693f00a914dc543bfd2&run=1

addyosmani commented on site/_apps/zuix-hn.txt:
```diff
@@ -6,19 +6,19 @@ library: 'ZUIX'
-lighthouse-link: https://www.webpagetest.org/lighthouse.php?test=170613_Y4_3f81ed03de518aaecc6a8823e74ec401&run=3
-wpt-em-link: https://www.webpagetest.org/result/170613_22_be140a755468b49f0e8480eeb79149ed/
-wpt-faster-3g-link: https://www.webpagetest.org/result/170613_Y4_3f81ed03de518aaecc6a8823e74ec401/
+interactive-em: '4.8s'
```
Can we update the numbers to the latest reports?

genemars replied:
Thanks @addyosmani for reviewing. I updated the numbers but also re-run the Faster-3G test since it looked strange that EM had a lower interactive-time. 
https://www.webpagetest.org/result/170715_PW_749f08a967caa216ca8cd749346601cb/

genemars commented:
@addyosmani ping :)

addyosmani commented:
Thanks for the ping :)

addyosmani commented:
@genemars Just to check, on repeat visits when offline I get served the application shell from service worker but not that JSON entries for the stories:

![image](https://user-images.githubusercontent.com/110953/28449079-c1cf0de2-6d91-11e7-83a7-bb0b1b8b379b.png)

Known bug/limitation of the implementation?

genemars commented:
@addyosmani it was a bug :smiley: 
WB caching route was set to `https://hacker-news.firebaseio.com/v0/item/(.*)`, so it was caching stories, but not indexes. Fixed it. Thanks.

addyosmani commented:
This looks good from my end. @anubhav7495 @kristoferbaxter any further suggestions from your side before we land this?

Very very minor things (which wouldn't block this landing):

* As top stories don't appear to be rendered in sequence, the population of them in the list UI is a little different to other apps. 
* "1 points" in news uses "points" plural instead of the singular (1 point)

kristoferbaxter commented:
Since there has been quite a few changes since the Lighthouse tests were performed on WPT, I ran another series. Could you update with the new times listed below?

Agree with @addyosmani on minor suggestions.

EM_3G, **6.2s**
https://www.webpagetest.org/result/170724_F6_c37ad291bb720b8c203da685d8a704b4/

Faster 3G, **4.8s**
https://www.webpagetest.org/result/170724_JT_fc286336283f4594b5e9a14ee486d437/

kristoferbaxter commented:
LGTM

genemars commented:
@addyosmani @kristoferbaxter @housseindjirdeh I've updated the results with @kristoferbaxter tests.
I've also fixed the "1 points" issue.
The list rendering/population is different because the implementation it's more like an Android ListView (asyncronous loaded items via item adapter).

addyosmani commented:
Thanks for re-running the numbers @kristoferbaxter and for updating, @genemars!
